src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,419 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.lang.stable;
+
+import jdk.internal.lang.StableValue;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+import static jdk.internal.lang.stable.StableUtil.*;
+
+public final class StableValueImpl<V> implements StableValue<V> {
+
+    private static final long MUTEX_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "mutex");
+
+    private static final long VALUE_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "value");
+
+    private static final long STATE_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "state");
+
+    /**
+     * An internal mutex used rather than synchronizing on `this`. Lazily created.
+     * If `null`    , we have not entered a mutex section yet
+     * If TOMBSTONE , we do not need synchronization anymore (isSet() && isError() = true)
+     * if instanceof Throwable , records a previous error and, we do not need synchronization anymore
+     * otherwise    , a distinct synchronization object
+     */
+    private Object mutex;
+
+    /**
+     * If non-null, holds a set value
+     * If `null`  , may be unset or hold a set `null` value
+     */
+    @Stable
+    private V value;
+
+    /**
+     * If StableUtil.NOT_SET  , a value is not set
+     * If StableUtil.NON_NULL , a non-null value is set
+     * If StableUtil.NULL     , a `null` value is set
+     */
+    @Stable
+    private int state;
+
+    /**
+     * Indicates if a supplier is currently being invoked. Used to
+     * detect circular supplier invocations.
+     */
+    @Stable private boolean supplying;
+
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        int s;
+        return (s = state) == NON_NULL || s == NULL ||
+                (s = stateVolatile()) == NON_NULL || s == NULL;
+    }
+
+    @ForceInline
+    @Override
+    public boolean isError() {
+        return state == ERROR || stateVolatile() == ERROR;
+    }
+
+    @ForceInline
+    @Override
+    public V orThrow() {
+        // Optimistically try plain semantics first
+        final V v = value;
+        if (v != null) {
```

Comment:
If `value == null && state == NULL`, can the path still be constant folded? I doubt it because `value` in this case may not be promoted to constant.

Reply:
Maybe the `state == NULL` check should be moved before `v != null`, as the **JIT** doesn’t constant‑fold `null` [`@Stable`] values:
https://github.com/openjdk/jdk/blob/8a4315f833f3700075d65fae6bc566011c837c07/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L41-L44
https://github.com/openjdk/jdk/blob/8a4315f833f3700075d65fae6bc566011c837c07/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java#L64-L71

[`@Stable`]: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

Reply:
It seems reasonable to assume `null` values are not constant-folded. For straight-out-of-the-box usage, there is no apparent significant difference as indicated by a new benchmark I just added:

```
Benchmark                                  Mode  Cnt      Score      Error   Units
StableStaticBenchmark.atomic              thrpt   10   5729.683 ?  502.023  ops/us
StableStaticBenchmark.dcl                 thrpt   10   6069.222 ?  951.784  ops/us
StableStaticBenchmark.dclHolder           thrpt   10   5502.102 ? 1630.627  ops/us
StableStaticBenchmark.stable              thrpt   10  12737.158 ? 1746.456  ops/us <- Non-null benchmark
StableStaticBenchmark.stableHolder        thrpt   10  12053.978 ? 1421.527  ops/us
StableStaticBenchmark.stableList          thrpt   10  12443.870 ? 2084.607  ops/us
StableStaticBenchmark.stableNull          thrpt   10  13164.232 ?  591.284  ops/us <- Added null benchmark
StableStaticBenchmark.stableRecordHolder  thrpt   10  13638.893 ? 1250.895  ops/us
StableStaticBenchmark.staticCHI           thrpt   10  13639.220 ? 1190.922  ops/us```

If the `null` value participates in a much larger constant-folding tree, there might be a significant difference. I was afraid moving the order would have detrimental effects on instance performance but that does not seem to be the case:

Checking value first:

```
Benchmark                           Mode  Cnt     Score      Error   Units
StableBenchmark.atomic             thrpt   10   246.460 ?   75.417  ops/us
StableBenchmark.dcl                thrpt   10   243.481 ?   35.021  ops/us
StableBenchmark.stable             thrpt   10  4977.693 ?  675.926  ops/us
StableBenchmark.stableHoldingList  thrpt   10  3614.460 ?  275.140  ops/us
StableBenchmark.stableList         thrpt   10  3328.155 ?  898.202  ops/us
StableBenchmark.stableListStored   thrpt   10  3842.174 ?  535.902  ops/us
StableBenchmark.stableNull         thrpt   10  6217.737 ?  840.376  ops/us
StableBenchmark.supplier           thrpt   10  9369.934 ? 1449.182  ops/us
```

Checking null first:

```
Benchmark                           Mode  Cnt     Score      Error   Units
StableBenchmark.atomic             thrpt   10   275.952 ?   39.480  ops/us
StableBenchmark.dcl                thrpt   10   252.697 ?   18.645  ops/us
StableBenchmark.stable             thrpt   10  5211.552 ?  315.307  ops/us
StableBenchmark.stableHoldingList  thrpt   10  3764.202 ?  224.325  ops/us
StableBenchmark.stableList         thrpt   10  3689.870 ?  419.858  ops/us
StableBenchmark.stableListStored   thrpt   10  3676.182 ?  938.485  ops/us
StableBenchmark.stableNull         thrpt   10  6046.935 ? 1512.391  ops/us
StableBenchmark.supplier           thrpt   10  9202.202 ? 1479.950  ops/us
```

So, swapping order seams to be the right move.

Reply:
I think the result would be more convincing if the stable case is changed to `sum += (stable.orThrow() == null ? 0 : 1) + (stable2.orThrow() == null ? 0 : 1);` as adding by 1 might be somewhat better optimized by JIT.

Reply:
I have instead changed parts of the `stableNull()` body to:

```
sum += (stableNull.orThrow() == null ? VALUE : VALUE2) + (stableNull2.orThrow() == null ? VALUE : VALUE2);
```